server/src/main/java/io/crate/execution/dml/Indexer.java

```diff
@@ -661,15 +665,40 @@ public ParsedDocument index(IndexItem item) throws IOException {
         }
     }
 
+    /**
+     *
+     * RETURNING can refer to multiple non-deterministic columns,
+     * generated and defaults and their combinations, e.g., gen_col + def_col.
+     *
+     * We need to ensure that we use already computed values for such references,
+     * this this must be called after {@link #index(IndexItem)} have been called.
+     *
+     */
     @Nullable
     public Object[] returnValues(IndexItem item) {
-        if (this.returnValueInputs == null) {
+        if (this.returnValues == null) {
             return null;
         }
+
+        for (int i = 0; i < returnValues.length; i++) {
+            Symbol updatedReturnValue = RefReplacer.replaceRefs(returnValues[i], ref -> {
+                if (undeterministicIdents.contains(ref.column())) {
+                    Object value = undeterministicValues.get(ref.column());
+                    return Literal.ofUnchecked(ref.valueType(), value);
+                }
```

Comment:
We are just using stored computed values now.

In `legacyInsert` (which is removed now) we [passed source](https://github.com/crate/crate/commit/672a882a82a442cdb007fdfe7eac9391d17cf5e6#diff-afc8b63b41c53d6e6ea5b956bdd39557ab55e1dfe1b3e1242f1d4e29a9fb57a0L607) (Doc) and parsed it via [DocRefResolver expressions](https://github.com/crate/crate/commit/672a882a82a442cdb007fdfe7eac9391d17cf5e6#diff-ec600b9f2d9172b3dd21c8acc88623c601fa3cc4db2ed3986682b35308024089L45) so this is still lightweight.